### ISSUE TITLE
Charset and Collation across databases

### ISSUE THREAD

**Issue by rijnhard** [doctrine/dbal#2351]
Fiddling with `collate` and `charset` table options could have unintended side effects. 
When making queries that cross between databases/tables with different collations 

e.g. in MySql

> Illegal mix of collations (latin1_general_cs,IMPLICIT) and (latin1_general_ci,IMPLICIT) for operation '='

There are a number of ways around this, but at present doctrine doesn't make it easy. 

Whats needed is a few additional functions:
- get the default Collation and Charset for a database
- get the default Collation and Charset for a table
- get the Collation and Charset for a column

Ideally it should be able to determine when 2 tables are being used a query and they have incompatible collations, then it should define a common collation or collate individual columns. 

I'm not sure what the best approach is, a little bit out of my depth here. But I will be implementing functions to do some of the above in my own project and wouldn't mind contributing.

The use case I have in particular is that I need to draw up a view between a table on a db over which I have no control (`DB1.T1`) and then a database where I have my entities generated to (`DB2.entity1`). In this particular case I am using the dbal's `QueryBuilder` to draw up the query for the `View` and one of the limitations is that I can't specify a collation in the QueryBuilder, a fairly nasty way around that is to manipulate the sql string before passing it to the `View`. But this still doesn't solve the problem of finding what the collation and charset should be set to.

Basically in my use case the question becomes: should I create the entities with the same charset/collation as `DB1.T1` or should I collate in the queries between the tables.

**Comment by rijnhard** [comment]
I have a working implementation for mysql that I have made to resemble how doctrines design (abstractions, function names and all)

I am still debating the actual design. But I have no problem separating it and actually applying it to the dbal code for a PR, I don't mind implementing the other databases either (as much as I can).

So far I ended up implementing `support` and `get___Sql` functions for encoding and charset on database level, table level, and column level. 

At the moment i only use Table level. 

In my particular use case I made an `EventListener` for doctrines `EventManager` that checked for a `MatchEncoding` annotation that I supplied with `dbName, tableName, ColumnName=null` in order to have the table generate so that it matches the encoding on a read only table. 

Works like a charm. Even though I still feel that it is probably possible to determine this automatically, without the users even knowing it's happening.
